src/runtime_src/core/pcie/driver/linux/xocl/subdev/hwmon_sdm.c

```diff
@@ -275,6 +276,151 @@ static int get_sensors_data_by_sensor_id(struct platform_device *pdev,
+	int buf_index, rcvd_rid, i, sz = 0;
+	char *cu_fmt = "%s,%u,%u,%u,%u\n";
+
+	buf_index = SDR_COMPLETE_IDX;
+	completion_code = in_buf[buf_index];
+	if(completion_code != SDR_CODE_OP_SUCCESS) {
+		dump_error_message(sdm, completion_code);
+		return -EINVAL;
+	}
+
+	buf_index = SDR_REPO_IDX;
+	repo_type = in_buf[buf_index];
+	rcvd_rid = sdr_get_id(repo_type);
+	if ((rcvd_rid < 0) || (rcvd_rid != repo_id) ||
+		(repo_id >= XGQ_CMD_SENSOR_SID_MAX)) {
+		xocl_err(&sdm->pdev->dev, "SDR Responce has invalid REPO TYPE: %d", repo_type);
+		return -EINVAL;
+	}
+
+	buf_index = buf_index + 1;
+	buf_len = in_buf[buf_index] + 3;
+
+	for (i = 0; i < sid_len; i++) {
+		buf_index = SDM_BUF_IDX_INCR(buf_index, 1, buf_len);
+		if (buf_index < 0)
+			goto abort;
+		val_len = in_buf[buf_index];
+
+		buf_index = SDM_BUF_IDX_INCR(buf_index, 1, buf_len);
+		if (buf_index < 0)
+			goto abort;
+		memcpy(&ins_val, &in_buf[buf_index], val_len);
+		sdm->sinfo[repo_id][i].value = ins_val;
+
+		buf_index = SDM_BUF_IDX_INCR(buf_index, val_len, buf_len);
+		if (buf_index < 0)
+			goto abort;
+		memcpy(&max_val, &in_buf[buf_index], val_len);
+		sdm->sinfo[repo_id][i].max = max_val;
+
+		buf_index = SDM_BUF_IDX_INCR(buf_index, val_len, buf_len);
+		if (buf_index < 0)
+			goto abort;
+		memcpy(&avg_val, &in_buf[buf_index], val_len);
+		sdm->sinfo[repo_id][i].avg = avg_val;
+
+		buf_index = SDM_BUF_IDX_INCR(buf_index, val_len, buf_len);
+		if (buf_index < 0)
+			goto abort;
+		status = in_buf[buf_index];
+		sdm->sinfo[repo_id][i].status = status;
+
+		sz += scnprintf(buf + sz, PAGE_SIZE - sz, cu_fmt,
+                        sdm->sinfo[repo_id][i].name,
+                        ins_val, avg_val, max_val, status);
+	}
+
+abort:
+	return sz;
+}
+
+static int show_sensors_raw(struct xocl_hwmon_sdm *sdm, char *buf,
+                            uint8_t repo_id)
+{
+	xdev_handle_t xdev = xocl_get_xdev(sdm->pdev);
+	int ret = 0, kind;
+	size_t resp_len = RESP_LEN;
+	char* sdr_buf;
+	int repo_type;
+	uint64_t data_args = 0;
+
+	sdr_buf = vzalloc(resp_len);
+	if (!sdr_buf)
+		goto done;
+
+	if (sdm->privileged) {
+		ret = xocl_xgq_collect_all_inst_sensors(xdev, sdr_buf, repo_id, RESP_LEN);
+	} else {
+		repo_type = to_sensor_repo_type(repo_id);
+		kind = to_xcl_sdr_type(repo_type);
+		if (kind < 0) {
+			xocl_err(&sdm->pdev->dev, "received invalid xcl grp type: %d", kind);
+			return -EINVAL;
+		}
+		data_args = 0x1 << MBREQ_INST_SENSORS_ENABLE_BIT;
+		ret = hwmon_sdm_read_from_peer(sdm->pdev, repo_type, kind, sdr_buf, resp_len, data_args);
+	}
+	if (!ret)
+		ret = parse_inst_sensors_info(sdm, sdr_buf, buf, repo_id);
+	else
+		xocl_err(&sdm->pdev->dev, "inst_sensor request for repo_id is failed with err: %d", ret);
+
+	vfree(sdr_buf);
+done:
+	return ret;
+}
+
+static ssize_t
+voltage_sensors_raw_show(struct device *dev, struct device_attribute *attr,
+                         char *buf)
+{
+	struct xocl_hwmon_sdm *sdm = dev_get_drvdata(dev);
+	ssize_t ret;
+
+	mutex_lock(&sdm->sdm_lock);
```

Comment:
can we move these mutex_lock/_unlock calls into show_sensors_raw() so that we don't have to repeat them in multiple locations?

Reply:
Hi Max, thanks for the review. I have taken care of your suggestions. Thank you.